gen_future_weather_file.py

```diff
@@ -0,0 +1,425 @@
+    # The headers dont need to be changed from here, however the data should
+    # be converted into its specific types.
+    with open(epw_file, "r") as epw:
+        reader = csv.reader(epw)  # Read the epw file as a csv
+        csv_data = list(reader)  # Convert the csv reader into a list.
+        data = csv_data[8:]  # Grab the data from the file (rows 8 and on.)
+        headers = csv_data[:8]  # Grab the headers from the file (rows 0 to 7.)
+
+    # Convert each cell in each row of data to its specific type. This way they
+    # can be changed as required by later functions.
+    for index, row in enumerate(data):
+
+        # Get the date of this row so that we can turn it into a datetime
+        # object for easier use later.
+        row_dates = [int(cell) for cell in row[:5]]
+        row_time = datetime(row_dates[0], row_dates[1],
+                            row_dates[2], row_dates[3] - 1,
+                            row_dates[4])
+
+        # Cells 6 to -9 (9 back from the end) can easilly be just sorted into
+        # ints or floats (float if there's a . in the string, int otherwise.)
+        # However, at position -9 we have a series of numbers representing
+        # weather codes. This must be left as a string.
+        row_vals = [float(c) if "." in c else int(c) for c in row[6:-9]]
+
+        row_vals.append(row[-8])  # Leave unchanged -8 as a string.
+
+        # The rest of the values in the file can be safely converted to ints or
+        # floats without loosing any key information.
+        for cell in row[-7:]:
+            if "." in cell:
+                row_vals.append(float(cell))
+            else:
+                row_vals.append(int(cell))
+
+        # Actually change the row we just worked on to have the values we
+        # created above.
+        data[index] = [row_time, row[5]]
+        for cell in row_vals:
+            data[index].append(cell)
+
+    # Return the data and the headers.
+    return data, headers
+
+
+def write_epw_data(data: list, headers: list, filename: str):
+    """ write_epw_data(list, list, str)
+
+        Combines the passed headers and data into an epw file with
+        the name filename.
+
+        Args:
+            data(list): A list of lists, each inner list is a row of
+                    data to be written to the epw file.
+            header(list): A list of lists, each inner list is a row of
+                      the header to be written to the epw file.
+            filename(str): The name of the file to be written.
+    """
+
+    # Reassemble the headers into one string. Each inner list of headers
+    # gets its own line, and each item in each inner list is comma seperated.
+    for index, header in enumerate(headers):
+        headers[index] = ",".join(header)
+    headers = "\n".join(headers)
+
+    # Rebuild the list of data into a string so that we can write it to a file.
+    epw_file = headers + "\n"
+    for data_row in data:
```

Comment:
Again, rather than editing the data that you're iterating over, and then building up a (presumably) lengthy variable `epw_file` that contains *all* of the output data, simply write it as you iterate.

A good way to handle this can be to write a delegate function that does all of the munging for one line, and another that does the output. So you can say:

~~~python
with open(filename, "w+") as epw:
    for row in data:
        output = munge_row(data)
        epw.write(output)
~~~